src/smt.rs

```diff
@@ -1,163 +1,632 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTSort {
+    Bool,
+    BV(u64),
+    Array(Box<SMTSort>, Box<SMTSort>),
+}
+
+pub fn bv(width: u64) -> SMTSort {
+    SMTSort::BV(width)
+}
+
+pub fn array(index: SMTSort, elem: SMTSort) -> SMTSort {
+    SMTSort::Array(Box::new(index), Box::new(elem))
+}
+
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct SMTVariable {
     pub name: String,
+    pub sort: SMTSort,
 }
 
-impl SMTFormat for SMTVariable {
-    fn as_smt(&self) -> String {
-        self.name.clone()
+impl SMTVariable {
+    pub fn new(name: String, sort: SMTSort) -> Self {
+        SMTVariable { name, sort }
     }
 }
 
-pub trait SMTFormat {
-    fn as_smt(&self) -> String;
+/// We keep the SMT expressions loose and runtime based for now.
+/// A potential future refactor may add compile time guarantees
+/// that illegal expressions cannot be built.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SMTExpr {
+    pub op: SMTOp,
+    pub args: Vec<SMTExpr>,
 }
 
-type BoxedSMTFormat = Box<dyn SMTFormat>;
+impl From<u64> for SMTExpr {
+    fn from(input: u64) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Literal(format!("{:064x}", input)),
```

Comment:
Why don't we add the `#x` prefix here right away?

Reply:
Because then we have to add that everywhere that creates a Literal, there are at least 2 more places. Eventually we should also make the printing Sort dependent.